custom_components/ocpp/api.py

```diff
@@ -1126,29 +1126,38 @@ def on_meter_values(self, connector_id: int, meter_value: dict, **kwargs):
                     measurand = DEFAULT_MEASURAND
                     unit = DEFAULT_ENERGY_UNIT
 
+                if measurand == DEFAULT_MEASURAND and unit is None:
+                    unit = DEFAULT_ENERGY_UNIT
+
+                if self._metrics[csess.meter_start.value].value == 0:
+                    # Charger reports Energy.Active.Import.Register directly as Session energy for transactions.
+                    self._charger_reports_session_energy = True
+
                 if phase is None:
                     if unit == DEFAULT_POWER_UNIT:
                         self._metrics[measurand].value = float(value) / 1000
                         self._metrics[measurand].unit = HA_POWER_UNIT
-                    elif unit == DEFAULT_ENERGY_UNIT or "Energy" in str(measurand):
-                        if self._metrics[csess.meter_start.value].value == 0:
-                            # Charger reports Energy.Active.Import.Register directly as Session energy for transactions
-                            self._charger_reports_session_energy = True
-                        if (
-                            transaction_matches
-                            and self._charger_reports_session_energy
-                            and measurand == DEFAULT_MEASURAND
-                            and connector_id
-                        ):
-                            self._metrics[csess.session_energy.value].value = (
-                                float(value) / 1000
+                    elif (
+                        measurand == DEFAULT_MEASURAND
+                        and self._charger_reports_session_energy
+                    ):
+                        if transaction_matches:
+                            if unit == DEFAULT_ENERGY_UNIT:
+                                value = float(value) / 1000
```

Comment:
Left it out because on line 366 in ChargePoint init, session_energy unit is set to "kWh": `self._metrics[csess.session_energy.value].unit = UnitOfMeasure.kwh.value`

So here only the value is changed to match the preset unit. Should this change be added though?

Reply:
Best to include for clarity and consistency imho